solveset modifications for use of Piecewise

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
fixes #14223 
fixes #10158
fixes #14238
closes #13913 as an alternative
do not address #10532 which is a WIP at #10751

#### Brief description of what is fixed or changed


Rather than modifying sovleset to recognize functions which can be written in terms of Piecewise, the `f` passed to `solveset` is now rewritten unconditionally in terms of Piecewise. In addition:

- Piecewise needed modification to pay attention to the given domain when computing the sets in which the expressions are valid
- solving of Relational (Eq and otherwise) was moved to `_solveset` since solution of Piecewise which uses `_solveset` can entail solving for expressions that are Relational (Eq and otherwise)
- the solution of 0 is excluded from `_invert` if it is not in the domain

#### Other comments

There is an asymmetry in how `solveset_real` and `invert_real` behave that I suspect exists because unlike `solveset`, the name `invert` was already taken by `sympy.polys.polytools.invert`. Whereas `solveset_real` doesn't accept a domain, `invert_real` does and simply passes it along to `_invert` so if one uses a domain other than S.Reals it doesn't matter if `invert_complex` (alias for `_invert`) or `invert_real` is called with that domain.

- [ ] check that the solution of Abs(Max(Abs(x-1),2-x))-3 is {-1, 4} and guard all Abs from replacements or perhaps only do so if the folded expression's symbol-dependent part is Abs. Maybe this for guarding Abs:

```
    mask = [Tuple(*i) for i in zip(ordered(f.atoms(Abs)), numbered_symbols('a', cls=Dummy)))]
    for i in range(len(mask)):
        for j in range(i + 1, len(mask)):
            mask[j] = mask[j].replace(*mask[i])
    for o, n in mask:
        f = f.replace(o, n)
    f = f.rewrite(Piecewise) # everything that's not in Abs
    for o, n in reversed(mask):
        # everything *in* abs
        f.replace(n, o.args[0].func(o.args[0].rewrite(Piecewise)))
    f = piecewise_fold(f)
```

smichr committed:
solveset attention to domain

In addition, Piecewise is solved only on S.Reals and
when as_expr_set_pairs is called, the domain is passed
to the returned intervals can be clipped according to the
domain.

_invert makes sure to use the domain even when the
solution is 0.

smichr committed:
use rewrite(Piecewise) in solveset

smichr committed:
move Relational solving to _solveset

Since Piecewise can contain Relationals we need to either
call solveset from _solveset or move the Relational solving
blocks into _solveset. The latter has been done.

When solving Piecewise, the conditions can only be solved
in the Real domain so the domain intersection with S.Reals
is passed to the as_expr_set_pairs function. The domain
itself is not changed, however, so if an attempt to solve
a Piecewise with expressions that are Relational is made an
error will be raised just like it would be raised if the
Relational didn't appear in a Piecewise:

>>> solveset(Lt(x, 2), x)
...
NotImplementedError:
Inequalities in the complex domain are not supported. Try the real
domain by setting domain=S.Reals

Yathartha22 commented:
> invert_real does and simply passes it along to _invert so if one uses a domain other than S.Reals it doesn't matter if invert_complex (alias for _invert) or invert_real is called with that domain.

+1 to the point you are trying to make, I think `invert_real` and `invert_complex` should just have `S.Reals` and `S.Complexes` as the domain respectively.

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -789,16 +785,29 @@ def _solveset(f, symbol, domain, _check=False):
+            f, symbol, domain=domain, relational=False)
+        except NotImplementedError:
+            result = ConditionSet(symbol, f, domain)
+        return result
```
Still can't get why these have to get here. Can you give an example which would have failed otherwise?

smichr replied:
Any Piecewise that has a Relational expression, e.g. `solveset(Piecewise((x<1,x<5),(x<2,True)),x,S.Reals)` or `solveset(Piecewise((Eq(x - 1, 0), x<0), (Eq(x - 4, 0), True)), x, S.Reals)`.

smichr committed:
watch for trivially different conj/arg

The Abs(r + Piecewise((0, r > 0), (1 - r, True))) changed
to Abs(r + Piecewise((0, r > 0), (-(r - 1), True))) and that
was the only difference between arg and conj. Now conj is
signsimped, too, so it appears the same.

Yathartha22 committed:
added checks for improper domain and symbols

Yathartha22 committed:
added tests

smichr committed:
rework masking

Yathartha22 commented:
@smichr this looks good as of now. I will have another look at it. Also, I have added a PR to your forked for improper parameters.

smichr committed:
additional test for non-Expr SymPy object

smichr commented:
Let me know when you've had a look and whether you would like to commit this.

Yathartha22 commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -988,9 +1000,8 @@ def test_solveset_domain():
     from sympy.solvers.solveset import _has_rational_power
     x = Symbol('x')
-    y = exp(x+1/x**2)
-    solution = solveset(y**2+y, x, S.Reals)
```
Interesting to see this test pass, just by making the symbol `real`.

Yathartha22 commented:
LGTM @smichr

Yathartha22 commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1676,10 +1687,43 @@ def test__is_finite_with_finite_vars():
+    assert solveset(x*Max(x, 15) - 10, x, dom) == FiniteSet(2/S(3))
+    assert solveset(x*Min(x, 15) - 10, x, dom) == FiniteSet(
+        -sqrt(10), sqrt(10))
+    # the following gets rewritten to
```
>  do you have any ideas about this situation?

What I figured out @smichr is that this equation gets converted to `Piecewise((x - 1, x + 2 >= Abs(x - 3) - 1), (Abs(x - 3) - 4, True))`, which further divides into `[(x - 1, Interval(0, oo)), (Abs(x - 3) - 4, Interval.open(-oo, 0))]`, now since both these `expr` are in the real domain, for the case of `Abs` _invert_real is called giving the result as {-1}. There is no `ValueError` because `_invert_real` solves it, while in cases for bare Abs in complex domain, there is no implementation in `_invert_complex` therfore they will always go to `_solve_abs` and raise error.

Yathartha22 replied:
> Also Max(x, y) is rewritten assuming that the args are real even though it is not defined for complex values. I'm not sure if that is consistent with Abs(x) not being rewritten unless x is real (but Abs is defined for complex values).

Yes, max or min is rewritten irrespective the args are real or complex [see here](https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/miscellaneous.py#L750), but in case of `Abs` it checks for args to be real, then only rewrites. [see here](https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/complexes.py#L586)

Two things need to be done:

- Restrict rewriting of Min/Max only to real,

- Also, Abs should be made to be able to rewrite in complex domain. Atleast solveset should solve Abs in complex domain (should be solved in another PR)

smichr replied:
It could be just as simple as only solving Piecewise if the conditions don't involve relationals when the variable is not real:
```
>>> solveset(x<1,x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\leslie\sympy\sympy\solvers\solveset.py", line 1006, in solveset

    setting domain=S.Reals'''))
NotImplementedError:
Inequalities in the complex domain are not supported. Try the real
domain by setting domain=S.Reals
>>> solveset(Eq(x,1),x)
{1}
>>> solveset(Eq(x,I),x)
{I}
```

Yathartha22 replied:
> It could be just as simple as only solving Piecewise if the conditions don't involve relationals when the variable is not real:

Rather than raising error for Piecewise let `_solve_abs` handle it. Committed in your repo.

smichr replied:
> If you agree with the above I would like to make the changes

Ok, I will raise an error if an attempt to solve a Piecewise with non-Equality conditions in the complex domain is made. Then this can be committed and your changes can be a separate PR. How does that sound?

Yathartha22 replied:
Sounds great!

smichr committed:
ValueError for solving Piecewise in complex domain
